resolver/resolver.cc

```diff
@@ -3435,18 +3435,30 @@ class ResolveSignaturesWalk {
             if (spec != sig.argTypes.end()) {
                 ENFORCE(spec->type != nullptr);
 
-                // It would be nice to remove the restriction on more than these two specific binds, but that would
-                // raise a lot more errors
-                if (!isSyntheticBlkArg && (spec->rebind == core::Symbols::MagicBindToAttachedClass() ||
-                                           spec->rebind == core::Symbols::MagicBindToSelfType())) {
-                    if (auto e = ctx.state.beginError(spec->loc, core::errors::Resolver::BindNonBlockParameter)) {
+                // TODO(#4095) Raise error if `bind` used for non-`&blk` arg
+                if (!isBlkArg && (spec->rebind == core::Symbols::MagicBindToAttachedClass() ||
+                                  spec->rebind == core::Symbols::MagicBindToSelfType())) {
+                    if (auto e = ctx.state.beginError(spec->nameLoc, core::errors::Resolver::BindNonBlockParameter)) {
                         e.setHeader("Using `{}` is not permitted here", "bind");
                         e.addErrorNote("Only block arguments can use `{}`", "bind");
                     }
                 }
 
                 arg.type = std::move(spec->type);
-                arg.loc = spec->loc;
+                if (isBlkArg && !core::Types::isSubType(ctx, arg.type, core::Types::nilableProcClass())) {
+                    if (auto e = ctx.state.beginError(spec->nameLoc, core::errors::Resolver::InvalidMethodSignature)) {
+                        e.setHeader("Block argument type must be either `{}` or a `{}` type (and possibly nilable)",
+                                    "Proc", "T.proc");
+                        e.addErrorNote(
+                            "Ruby allows classes to have custom `{}` methods to convert `{}`-passed arguments into\n"
+                            "    `{}` objects, but even those methods must return `{}` objects.",
+                            "to_proc", "&blk", "Proc", "Proc");
```

Comment:
OOC, do we have tests for honoring this magic `to_proc` behavior?

Reply:
sort of but they could be better. i added one.